File: Assets/CoachEngine/Scripts/NativeBridge/iOS/ComponentRuntime.cs
/*
 * Copyright (c) 2021 AGT Group (R&D) GmbH. All rights reserved.
 * Redistribution and use in source and binary form without the express
 * permission of the above copyright holder is not permitted.
 */

#if PLATFORM_IOS && !UNITY_EDITOR
using Newtonsoft.Json;
using UnityEngine;

namespace CoachAiEngine.Ios {
    public partial class ComponentRuntime : IComponentRuntime {
        private int _nativeRuntimeHandle;
        public bool Disposed => _nativeRuntimeHandle == -1;
        private readonly int _runtimeHandle;

        public ComponentRuntime(IComponentRuntimeCallbacks callbackHandler) {
            Debug.Log("Starting ComponentRuntime");
            _runtimeHandle = Registry.RegisterRuntime(callbackHandler);
            _nativeRuntimeHandle = coach_cr_create();
        }

        ~ComponentRuntime() => Stop();

        /**
         * <inheritdoc cref="IComponentRuntime.Load"/>
         */
        public void Load(CoachComponent component) {
            if (Disposed) return;
            var config = JsonConvert.SerializeObject(component.Parameters);
            coach_cr_load(_nativeRuntimeHandle, component.ComponentId, config, OnComponentLoaded, _runtimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Configure"/>
         */
        public void Configure(CoachComponent component) {
            if (Disposed) return;
            var runConfig = JsonConvert.SerializeObject(component.Parameters);
            coach_cr_configure(_nativeRuntimeHandle, component.ComponentId, runConfig, OnComponentConfigured, _runtimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Start"/>
         */
        public void Start() {
            if (Disposed) return;
            coach_cr_resume(_nativeRuntimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Pause"/>
         */
        public void Pause() {
            if (Disposed) return;
            coach_cr_pause(_nativeRuntimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Stop"/>
         */
        public void Stop() {
            if (Disposed) return;
            var handle = _nativeRuntimeHandle;
            _nativeRuntimeHandle = -1;
            coach_cr_dispose(handle);
            Registry.DeregisterRuntime(_runtimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Subscribe"/>
         */
        public long Subscribe(string eventId) {
            if (Disposed) {
                Debug.LogError($"{nameof(ComponentRuntime)}: runtime is disposed.");
                return -1;
            }
            return coach_cr_subscribe(_nativeRuntimeHandle, eventId, OnEvent, _runtimeHandle);
        }

        /**
         * <inheritdoc cref="IComponentRuntime.Unsubscribe"/>
         */
        public void Unsubscribe(long subscriptionId) {
            if (Disposed) return;
            coach_cr_unsubscribe(_nativeRuntimeHandle, subscriptionId);
        }
    }
}
#endif

File: Assets/Plugins/CoachEngine/Android/ComponentRuntime.kt
/*
 * Copyright (c) 2021 AGT Group (R&D) GmbH. All rights reserved.
 * Redistribution and use in source and binary form without the express
 * permission of the above copyright holder is not permitted.
 */

package com.coachai.engine.unity

import android.util.Log
import com.coachai.engine.ar.components.ARPublisher
import com.coachai.engine.ar.components.AugmentedRealitySourceParameter
import com.coachai.engine.camera.components.CameraPublisher
import com.coachai.engine.camera.components.VideoSourceParameter
import com.coachai.engine.concurrent.SubscriptionId
import com.coachai.engine.concurrent.dispatchqueue.DispatchQueue
import com.coachai.engine.recording.misc.serializeToJson
import com.coachai.engine.runtime.parameter.getIdentifier

@Suppress("UNUSED")
internal class ComponentRuntime(private val callbackHandler: IComponentRuntimeCallbacks) {
    private val queue = DispatchQueue.mainThreadQueue("UnityComponentRuntime")
    private val runtime = com.coachai.engine.components.ComponentRuntime()

    /**
     * Load the given component.
     */
    fun load(component: String, config: String) {
        Log.d(COACH_TAG, "Loading component $component")
        val parameters = parseParameterString(config).toMutableMap()

        when(component) {
            ARPublisher.identifier -> {
                parameters[AugmentedRealitySourceParameter.getIdentifier()] = CoachSdk.unitySession
            }
            CameraPublisher.identifier -> {
                parameters[VideoSourceParameter.getIdentifier()] = CoachSdk.unitySession
            }
        }

        queue.async {
            runtime.load(component, parameters) {
                Log.d(COACH_TAG, "Component $component loaded")
                callbackHandler.OnLoaded(component, true)
            }
        }
    }

    /**
     * Configure a loaded component.
     */
    fun reconfigure(component: String, config: String) {
        Log.d(COACH_TAG, "Configuring component $component")
        val parameters = parseParameterString(config)
        queue.async {
            runtime.reconfigure(component, parameters) {
                Log.d(COACH_TAG, "Component $component configured")
                callbackHandler.OnConfigured(component, true)
            }
        }
    }

    /**
     * Resume execution of loaded components.
     */
    fun start() = queue.sync {
        Log.d(COACH_TAG, "Resuming component runtime")
        runtime.start()
    }

    /**
     * Pause execution of loaded components.
     */
    fun pause() = queue.sync {
        Log.d(COACH_TAG, "Pausing component runtime")
        runtime.pause()
    }

    /**
     * Stop execution of all loaded components and dispose of any resources held by this runtime.
     */
    fun dispose() = queue.sync {
        Log.d(COACH_TAG, "Disposing component runtime")
        runtime.dispose()
    }

    /**
     * Subscribe to the given event.
     */
    fun subscribe(eventIdentifier: String) = runtime.subscribe(eventIdentifier) {
        val json = it.serializeToJson()
        callbackHandler.OnEvent(json)
    }

    /**
     * Cancel the subscription with the given id.
     */
    fun unsubscribe(subscriptionId: SubscriptionId) = runtime.unsubscribe(subscriptionId)

    private fun parseParameterString(json: String): Map<String, Any> = try {
        val jsonObject = org.json.JSONObject(json)
        jsonObject.keys().asSequence().associateWith { jsonObject[it] }
    } catch (_: org.json.JSONException) {
        Log.e(COACH_TAG, "Failed to parse json parameter string: $json.")
        emptyMap()
    }

    @Suppress("FunctionName")
    interface IComponentRuntimeCallbacks {
        fun OnLoaded(componentId: String, result: Boolean)
        fun OnConfigured(componentId: String, result: Boolean)
        fun OnEvent(event: String)
    }
}
